File: js/plugins/a2ui/tests/parser_test.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from 'node:assert';
import { describe, it } from 'node:test';
import { basicCatalog } from '../src/catalog.js';
import { A2uiStreamParser } from '../src/parser.js';
import type { A2uiEnvelope } from '../src/types.js';

function fixedId() {
  return 'surface-1';
}

/** Runs `fn` while capturing `console.warn` output, restoring it afterwards. */
function captureWarnings(fn: () => void): string[] {
  const warnings: string[] = [];
  const original = console.warn;
  console.warn = (...args: unknown[]) => {
    warnings.push(args.map(String).join(' '));
  };
  try {
    fn();
  } finally {
    console.warn = original;
  }
  return warnings;
}

function collect(parser: A2uiStreamParser, chunks: string[]) {
  let prose = '';
  const batches: A2uiEnvelope[][] = [];
  for (const c of chunks) {
    const r = parser.push(c);
    prose += r.prose;
    batches.push(...r.envelopeBatches);
  }
  const f = parser.flush();
  prose += f.prose;
  batches.push(...f.envelopeBatches);
  return { prose, batches };
}

const SAMPLE_BLOCK = `\`\`\`a2ui
[
  { "createSurface": { "surfaceId": "SURFACE_ID", "catalogId": "${basicCatalog.id}" } },
  { "updateComponents": { "surfaceId": "SURFACE_ID", "components": [
    { "id": "root", "component": "Text", "text": "hi" }
  ] } }
]
\`\`\`
`;

describe('A2uiStreamParser', () => {
  it('separates prose from a complete a2ui block', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const { prose, batches } = collect(parser, [
      'Here is the weather:\n',
      SAMPLE_BLOCK,
    ]);
    assert.match(prose, /Here is the weather/);
    assert.doesNotMatch(prose, /createSurface/);
    assert.strictEqual(batches.length, 1);
    assert.strictEqual(batches[0].length, 2);
  });

  it('substitutes SURFACE_ID placeholder with the generated id', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const { batches } = collect(parser, [SAMPLE_BLOCK]);
    const create = batches[0][0] as any;
    assert.strictEqual(create.createSurface.surfaceId, 'surface-1');
    const update = batches[0][1] as any;
    assert.strictEqual(update.updateComponents.surfaceId, 'surface-1');
  });

  it('stamps the protocol version', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
      version: 'v0.9',
    });
    const { batches } = collect(parser, [SAMPLE_BLOCK]);
    assert.strictEqual((batches[0][0] as any).version, 'v0.9');
  });

  it('handles a block split across many tiny chunks', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const chunks = SAMPLE_BLOCK.match(/.{1,3}/gs) ?? [];
    const { prose, batches } = collect(parser, ['prefix ', ...chunks]);
    assert.match(prose, /prefix/);
    assert.strictEqual(batches.length, 1);
    assert.strictEqual(batches[0].length, 2);
  });

  it('does not leak a partial fence into prose', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    // Push text ending in a partial fence, then complete it.
    const r1 = parser.push('hello ```a2');
    assert.doesNotMatch(r1.prose, /```a2/);
    const { batches } = collect(parser, [
      'ui\n[{"createSurface":{"surfaceId":"SURFACE_ID","catalogId":"' +
        basicCatalog.id +
        '"}}]\n```\n',
    ]);
    assert.strictEqual(batches.length, 1);
  });

  it('emits prose with no blocks unchanged', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const { prose, batches } = collect(parser, ['just ', 'text ', 'here']);
    assert.strictEqual(prose, 'just text here');
    assert.strictEqual(batches.length, 0);
  });

  it('throws in strict mode on unknown component', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
      validate: 'strict',
    });
    const bad = `\`\`\`a2ui
[{ "updateComponents": { "surfaceId": "SURFACE_ID", "components": [
  { "id": "root", "component": "NotAThing" }
] } }]
\`\`\`
`;
    assert.throws(() => collect(parser, [bad]), /not in catalog/);
  });

  it('throws in strict mode when root is missing', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
      validate: 'strict',
    });
    const bad = `\`\`\`a2ui
[{ "updateComponents": { "surfaceId": "SURFACE_ID", "components": [
  { "id": "x", "component": "Text", "text": "hi" }
] } }]
\`\`\`
`;
    assert.throws(() => collect(parser, [bad]), /root/);
  });

  it('validate:off does not throw on bad JSON', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
      validate: 'off',
    });
    const bad = '```a2ui\n{not json}\n```\n';
    const { batches } = collect(parser, [bad]);
    assert.strictEqual(batches.length, 0);
  });

  it('validate:warn drops an unknown component without throwing', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
      validate: 'warn',
    });
    const bad = `\`\`\`a2ui
[{ "updateComponents": { "surfaceId": "SURFACE_ID", "components": [
  { "id": "root", "component": "NotAThing" }
] } }]
\`\`\`
`;
    const warnings = captureWarnings(() => {
      const { batches } = collect(parser, [bad]);
      assert.strictEqual(batches.length, 0);
    });
    assert.ok(
      warnings.some((w) => /not in catalog/.test(w)),
      'expected a warning about the unknown component'
    );
  });

  it('validate:warn drops bad JSON without throwing', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
      validate: 'warn',
    });
    const bad = '```a2ui\n{not json}\n```\n';
    const warnings = captureWarnings(() => {
      const { batches } = collect(parser, [bad]);
      assert.strictEqual(batches.length, 0);
    });
    assert.ok(
      warnings.some((w) => /JSON/.test(w)),
      'expected a warning about the malformed JSON'
    );
  });

  it('prepends a createSurface when a block only has updates', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const updateOnly = `\`\`\`a2ui
[{ "updateComponents": { "surfaceId": "SURFACE_ID", "components": [
  { "id": "root", "component": "Text", "text": "refreshed" }
] } }]
\`\`\`
`;
    const { batches } = collect(parser, [updateOnly]);
    assert.strictEqual(batches.length, 1);
    const first = batches[0][0] as any;
    assert.ok(
      first.createSurface,
      'expected a synthesized createSurface first'
    );
    assert.strictEqual(first.createSurface.surfaceId, 'surface-1');
    assert.strictEqual(first.createSurface.catalogId, basicCatalog.id);
    // The update targets the same surface id.
    const update = batches[0][1] as any;
    assert.strictEqual(update.updateComponents.surfaceId, 'surface-1');
  });

  it('does not add a second createSurface when one is present', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const { batches } = collect(parser, [SAMPLE_BLOCK]);
    const createCount = batches[0].filter((e: any) => e.createSurface).length;
    assert.strictEqual(createCount, 1);
  });

  it('handles two separate blocks in one turn', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    const { batches } = collect(parser, [
      SAMPLE_BLOCK,
      'some text between\n',
      SAMPLE_BLOCK,
    ]);
    assert.strictEqual(batches.length, 2);
  });

  it('preserves prose/block order in segments (prose after a block)', () => {
    const parser = new A2uiStreamParser({
      catalog: basicCatalog,
      surfaceId: fixedId,
    });
    // A single push containing: intro prose, a block, then trailing prose.
    const r = parser.push('before\n' + SAMPLE_BLOCK + 'after');
    const f = parser.flush();
    const segments = [...r.segments, ...f.segments];
    // Expect order: prose("before"), envelopes, prose("after").
    assert.strictEqual(segments.length, 3);
    assert.ok('prose' in segments[0] && /before/.test(segments[0].prose));
    assert.ok('envelopes' in segments[1]);
    assert.ok('prose' in segments[2] && /after/.test(segments[2].prose));
  });
});

File: go/plugins/a2ui/parser_test.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package a2ui

import (
	"strings"
	"testing"
)

// fixedSurfaceID returns a surface-id factory yielding a constant id.
func fixedSurfaceID(id string) func() string {
	return func() string { return id }
}

// collect concatenates prose and returns all envelope batches from segments.
func collect(segs []segment) (prose string, batches [][]Envelope) {
	for _, s := range segs {
		if s.isEnvelope {
			batches = append(batches, s.envelopes)
		} else {
			prose += s.prose
		}
	}
	return prose, batches
}

func TestParserProseOnly(t *testing.T) {
	p := newStreamParser(parserOptions{surfaceID: fixedSurfaceID("s1")})
	segs, err := p.push("Hello, world.")
	if err != nil {
		t.Fatal(err)
	}
	flushed, err := p.flush()
	if err != nil {
		t.Fatal(err)
	}
	prose, batches := collect(append(segs, flushed...))
	if prose != "Hello, world." {
		t.Errorf("prose = %q, want %q", prose, "Hello, world.")
	}
	if len(batches) != 0 {
		t.Errorf("got %d batches, want 0", len(batches))
	}
}

func TestParserExtractsBlock(t *testing.T) {
	catalog := BasicCatalog()
	p := newStreamParser(parserOptions{
		catalog:   catalog,
		validate:  ValidateWarn,
		surfaceID: fixedSurfaceID("s1"),
	})
	input := "Here is a card:\n```a2ui\n" +
		`[{"createSurface":{"surfaceId":"SURFACE_ID","catalogId":"` + catalog.ID + `"}},` +
		`{"updateComponents":{"surfaceId":"SURFACE_ID","components":[{"id":"root","component":"Text","text":"hi"}]}}]` +
		"\n```\nDone."
	segs, err := p.push(input)
	if err != nil {
		t.Fatal(err)
	}
	flushed, err := p.flush()
	if err != nil {
		t.Fatal(err)
	}
	prose, batches := collect(append(segs, flushed...))
	if !strings.Contains(prose, "Here is a card:") || !strings.Contains(prose, "Done.") {
		t.Errorf("prose = %q, missing expected text", prose)
	}
	if len(batches) != 1 {
		t.Fatalf("got %d batches, want 1", len(batches))
	}
	batch := batches[0]
	if len(batch) != 2 {
		t.Fatalf("got %d envelopes, want 2", len(batch))
	}
	cs, _ := batch[0]["createSurface"].(map[string]any)
	if cs["surfaceId"] != "s1" {
		t.Errorf("surfaceId = %v, want s1 (placeholder should be replaced)", cs["surfaceId"])
	}
}

func TestParserInjectsCreateSurface(t *testing.T) {
	catalog := BasicCatalog()
	p := newStreamParser(parserOptions{
		catalog:   catalog,
		validate:  ValidateWarn,
		surfaceID: fixedSurfaceID("s9"),
	})
	// Only an updateComponents; the parser must prepend a createSurface.
	input := "```a2ui\n" +
		`[{"updateComponents":{"surfaceId":"SURFACE_ID","components":[{"id":"root","component":"Text","text":"hi"}]}}]` +
		"\n```"
	segs, err := p.push(input)
	if err != nil {
		t.Fatal(err)
	}
	flushed, _ := p.flush()
	_, batches := collect(append(segs, flushed...))
	if len(batches) != 1 {
		t.Fatalf("got %d batches, want 1", len(batches))
	}
	batch := batches[0]
	if len(batch) != 2 {
		t.Fatalf("got %d envelopes, want 2 (createSurface injected)", len(batch))
	}
	if _, ok := batch[0]["createSurface"]; !ok {
		t.Errorf("first envelope should be createSurface, got %v", batch[0])
	}
}

func TestParserSplitAcrossChunks(t *testing.T) {
	catalog := BasicCatalog()
	p := newStreamParser(parserOptions{
		catalog:   catalog,
		validate:  ValidateWarn,
		surfaceID: fixedSurfaceID("s1"),
	})
	full := "prefix ```a2ui\n" +
		`[{"createSurface":{"surfaceId":"SURFACE_ID","catalogId":"c"}},` +
		`{"updateComponents":{"surfaceId":"SURFACE_ID","components":[{"id":"root","component":"Text","text":"hi"}]}}]` +
		"\n``` suffix"

	var allSegs []segment
	// Feed one rune at a time to stress the incremental fence handling.
	for _, r := range full {
		segs, err := p.push(string(r))
		if err != nil {
			t.Fatal(err)
		}
		allSegs = append(allSegs, segs...)
	}
	flushed, _ := p.flush()
	allSegs = append(allSegs, flushed...)

	prose, batches := collect(allSegs)
	if !strings.Contains(prose, "prefix") || !strings.Contains(prose, "suffix") {
		t.Errorf("prose = %q, missing prefix/suffix", prose)
	}
	if len(batches) != 1 {
		t.Fatalf("got %d batches, want 1", len(batches))
	}
	if len(batches[0]) != 2 {
		t.Errorf("got %d envelopes, want 2", len(batches[0]))
	}
}

func TestParserStrictRejectsUnknownComponent(t *testing.T) {
	catalog := BasicCatalog()
	p := newStreamParser(parserOptions{
		catalog:   catalog,
		validate:  ValidateStrict,
		surfaceID: fixedSurfaceID("s1"),
	})
	input := "```a2ui\n" +
		`[{"updateComponents":{"surfaceId":"SURFACE_ID","components":[{"id":"root","component":"NoSuchThing"}]}}]` +
		"\n```"
	_, err := p.push(input)
	if err == nil {
		t.Fatal("expected error for unknown component in strict mode")
	}
	if !strings.Contains(err.Error(), "NoSuchThing") {
		t.Errorf("error = %v, want mention of NoSuchThing", err)
	}
}

func TestParserWarnDropsBadJSON(t *testing.T) {
	catalog := BasicCatalog()
	p := newStreamParser(parserOptions{
		catalog:   catalog,
		validate:  ValidateWarn,
		surfaceID: fixedSurfaceID("s1"),
	})
	input := "```a2ui\nnot valid json\n```"
	segs, err := p.push(input)
	if err != nil {
		t.Fatalf("warn mode should not error, got %v", err)
	}
	flushed, _ := p.flush()
	_, batches := collect(append(segs, flushed...))
	if len(batches) != 0 {
		t.Errorf("got %d batches, want 0 (bad JSON dropped)", len(batches))
	}
}

func TestParserStrictErrorsBadJSON(t *testing.T) {
	p := newStreamParser(parserOptions{
		catalog:   BasicCatalog(),
		validate:  ValidateStrict,
		surfaceID: fixedSurfaceID("s1"),
	})
	_, err := p.push("```a2ui\n{ not json }\n```")
	if err == nil {
		t.Fatal("expected error for bad JSON in strict mode")
	}
}

func TestParserMissingRootRejected(t *testing.T) {
	p := newStreamParser(parserOptions{
		catalog:   BasicCatalog(),
		validate:  ValidateStrict,
		surfaceID: fixedSurfaceID("s1"),
	})
	input := "```a2ui\n" +
		`[{"updateComponents":{"surfaceId":"SURFACE_ID","components":[{"id":"body","component":"Text","text":"hi"}]}}]` +
		"\n```"
	_, err := p.push(input)
	if err == nil || !strings.Contains(err.Error(), "root") {
		t.Fatalf("expected root-required error, got %v", err)
	}
}
